packages/lit-html/src/test/directives/class-map_test.ts

```diff
@@ -0,0 +1,144 @@
+
+import {html, svg, render} from '../../lib/lit-html.js';
+import {ClassInfo, classMap} from '../../lib/directives/class-map.js';
+import {assert} from '@esm-bundle/chai';
+
+suite('classMap directive', () => {
+  let container: HTMLDivElement;
+
+  function renderClassMap(cssInfo: ClassInfo) {
+    render(html`<div class="${classMap(cssInfo)}"></div>`, container);
+  }
+
+  function renderClassMapStatic(cssInfo: ClassInfo) {
+    render(html`<div class="aa ${classMap(cssInfo)}  bb "></div>`, container);
+  }
+
+  setup(() => {
+    container = document.createElement('div');
+  });
+
+  test('adds classes', () => {
+    renderClassMap({foo: 0, bar: true, zonk: true});
+    const el = container.firstElementChild!;
+    assert.isFalse(el.classList.contains('foo'));
+    assert.isTrue(el.classList.contains('bar'));
+    assert.isTrue(el.classList.contains('zonk'));
+  });
+
+  test('removes classes', () => {
+    renderClassMap({foo: true, bar: true, baz: true});
+    const el = container.firstElementChild!;
+    assert.isTrue(el.classList.contains('foo'));
+    assert.isTrue(el.classList.contains('bar'));
+    assert.isTrue(el.classList.contains('baz'));
+    renderClassMap({foo: false, bar: true, baz: false});
+    assert.isFalse(el.classList.contains('foo'));
+    assert.isTrue(el.classList.contains('bar'));
+    assert.isFalse(el.classList.contains('baz'));
+  });
+
+  test('removes omitted classes', () => {
+    renderClassMap({foo: true, bar: true, baz: true});
+    const el = container.firstElementChild!;
+    assert.isTrue(el.classList.contains('foo'));
+    assert.isTrue(el.classList.contains('bar'));
+    assert.isTrue(el.classList.contains('baz'));
+    renderClassMap({});
+    assert.isFalse(el.classList.contains('foo'));
+    assert.isFalse(el.classList.contains('bar'));
+    assert.isFalse(el.classList.contains('baz'));
+    assert.equal(el.classList.length, 0);
+  });
+
+  test('works with static classes', () => {
+    renderClassMapStatic({foo: true});
+    const el = container.firstElementChild!;
+    assert.isTrue(el.classList.contains('aa'), 'aa 1');
+    assert.isTrue(el.classList.contains('bb'), 'bb 1');
+    assert.isTrue(el.classList.contains('foo'), 'foo 1');
+    renderClassMapStatic({});
+    assert.isTrue(el.classList.contains('aa'), 'aa');
+    assert.isTrue(el.classList.contains('bb'), 'bb');
+    assert.isFalse(el.classList.contains('foo'), 'foo');
+  });
+
+  test('works with imperatively added classes', () => {
+    renderClassMap({foo: true});
+    const el = container.firstElementChild!;
+    assert.isTrue(el.classList.contains('foo'));
+
+    el.classList.add('bar');
+    assert.isTrue(el.classList.contains('bar'));
+
+    renderClassMap({foo: false});
+    assert.isFalse(el.classList.contains('foo'));
+    assert.isTrue(el.classList.contains('bar'));
+  });
+
+  test('can not override static classes', () => {
+    renderClassMapStatic({aa: false, bb: true});
```

Comment:
Sorry, I forgot the behavior here. I had thought that statics could be overridden, but provided a default if they weren't in the dynamic class info. I think this is from @sorvell's initial implementation.

I guess there's two ways to look at static + dynamic classes. One is that statics provide the defaults, to be overridden by dynamic. The other is that dynamic classes are computed separately and after being determined are merged with static classes, so static classes can never be removed. That's what we have now and outside compelling reasons otherwise, I don't think we should make a breaking change here.